articles/service-fabric/service-fabric-work-with-reliable-collections.md

```diff
@@ -213,10 +213,48 @@ Internally, Reliable Collections serialize your objects using .NET's DataContrac
 
 > [!WARNING]
-> While you can modify the schema of a key, you must ensure that your key's hash code and equals algorithms are stable. If you change how either of these algorithms operate, you will not be able to look up the key within the reliable dictionary ever again.
+> While you can modify the schema of a key, you must ensure that your key's equality and comparison algorithms are stable.
+> Behavior of reliable collections after a change in either of these algorithms is undefined and may lead to data corruption, loss and
+> service crashes.
 > .NET Strings can be used as a key but use the string itself as the key--do not use the result of String.GetHashCode as the key.
 
-Alternatively, you can perform what is typically referred to as a two upgrade. With a two-phase upgrade, you upgrade your service from V1 to V2: V2 contains the code that knows how to deal with the new schema change but this code doesn't execute. When the V2 code reads V1 data, it operates on it and writes V1 data. Then, after the upgrade is complete across all upgrade domains, you can somehow signal to the running V2 instances that the upgrade is complete. (One way to signal this is to roll out a configuration upgrade; this is what makes this a two-phase upgrade.) Now, the V2 instances can read V1 data, convert it to V2 data, operate on it, and write it out as V2 data. When other instances read V2 data, they do not need to convert it, they just operate on it, and write out V2 data.
+Alternatively, you can perform a multi-phase upgrade. 
+1. Upgrade service to a new version that
+    - has both the original V1, and the new V2 version of the data contracts included in the service code package;
+    - registers custom V2 [state serializers](/azure/service-fabric/service-fabric-reliable-services-reliable-collections-serialization#custom-serialization), if needed;
+    - performs all operations on the original, V1 collection using the V1 data contracts.
+2. Upgrade service to a new version that
+    - [creates a new, V2 collection](/dotnet/api/microsoft.servicefabric.data.ireliablestatemanager.getoraddasync);
+    - performs each add, update and delete operation on first V1 and then V2 collections in a single transaction;
+    - performs read operations on the V1 collection only.
+3. Copy all data from the V1 collection to the V2 collection.
+    - This can be done in a background process by the service version deployed in step 2.
+    - [Enumerate all keys](/dotnet/api/microsoft.servicefabric.data.collections.ireliabledictionary2-2.createkeyenumerableasync)
+      in the V1 collection with the
+      [IsolationLevel.Snapshot](/dotnet/api/microsoft.servicefabric.data.beta.ireliablestatemanager2.createtransaction)
+      to avoid locking it for the entire duration of the copy process.
+    - For each key, use a separate transaction with the [IsolationLevel.ReadRepeatable](/dotnet/api/microsoft.servicefabric.data.beta.isolationlevel) to 

```

Comment:
I don't understand, would this be a nested transaction? That wouldn't be supported. Are we enumerating all the keys into memory in one transaction, then after that transaction completes, entering a for loop which enumerates that in memory list, trying to update each key in separate transactions one by one?

Reply:
OK, I understand. Maybe we can clarify the first bullet point by saying "Enumerate and load all keys into memory in a snapshot transaction".

And the next bullet point as "Using the list of keys, in a new transaction,..."

Reply:
"to avoid locking it for the entire duration of the copy process." -- Enumerate to get all the keys would not hold locks for the entire copy process.

Enumerations are always Snapshot anyway (even if CreateTxn specifies ReadRepeatable) so specifying it is not necessary (at least today).